### PR TITLE
Skeleton Implementation of DistributedSpectator at par with DistributedMode Controller cluster

### PR DESCRIPTION
**Pull request by gopalrajpurohit**
This is a model implementation of distributed spectator. (The Actual ConfigGenerator code has been placeholder code and not bound in this implementation. We will do it in next PR)

Why this change:
The current implementation of Spectator works in two different ways.
1. Runs as a standalone cluster using distributed lock using curator framework. The leader spectator looses leadership when it releases a lock due to issue with zk connections or when the process goes down. In case of when the Spectator goes down, it relies on an external restart mechanism to come back online. In case of zk connection reset, it does a system exist and then relies on external restart mechanism. The bigger downside though is that we need to allocate a cluster of Spectator instances (potentially 2-3 instances) per Participant cluster. We may have 100s of clusters, in which case this approach is wasteful of resources (as only one of the Spectator is really doing any work, others are sitting idle). We also need to maintain one Spectator cluster per Participant cluster. This adds to overall maintenance overhead.

2. Second method of running Spectator is from within a Participant cluster. In this case, the business logic has no way to know if a specific instance of Spectator is leader spectator unless it receives notification from helix. However there is no way to tell when the Spectator instance embedded in the Participant loses leadership, since there is only one callback method that client needs to implement which provides notification based on specific event types only. This approach is also not efficient for big clusters, where spectator (which is pretty much almost-stateless) eats away critical system resources from participant cluster's individual instance's resources. Even though the Spectator's core function (i.e. generation of config map) runs on single Participant node, we need to allocate enough JVM memory in each of the Participant node, which otherwise could be used in Participant. 
3. Managing the leadership state has also been buggy, as we recently found that multiple spectator instances could potentially be running and generating shard_map config. 

This distributed implementation (Currently in this PR, it is just a placeholder and does nothing other than connecting to the cluster) comes with a way manage custom states, in independent cluster. It is fully featured implementation that is at par with helix's own Distributed Controller implementation. In fact quite a bit of code has been taken from Distributed Controller implementation and modified to run custom code, instead of controller. This provides us an easy way of

2. Separating Participant clusters and Spectator cluster and enforcing separation of concern in a nit and clean way.

2. Saves resources in Spectator clusters, when we have multitude of Participant clusters. Since spectator cluster can handle multiple participant clusters, similar to controller cluster. Moreover adding another Participant cluster into existing spectator cluster will be much easy.. similar to how we add a participant to controller cluster, same mechanism will work for this implementation of Spectator in distributed mode.

3. Saves memory in big Participant clusters by not having to allocate memory that otherwise need to be taken up by Spectator core logic. This is especially useful, since during deploys, the Spectator may receive external views at much more frequency than during normal times. Same thing will happen during heavy rebalancing activities. By separating Spectator cluster from Participant, we allow all resources to be solely used by Participant.

4. Provides callback notifications to custom code, such that the business logic understands and knows when it is leader and when it is not.

5. Provides a mechanism to plug any other implementation by simply providing an implementation of  ClusterSpectatorFactory and binding custom implementation instead of provided ones. This allows one to use different mechanisms for distributing shard config, or even different format for shard generation.

### CODE REVIEW

File: cluster_management/src/main/java/com/pinterest/rocksplicator/spectator/ConfigGeneratorClusterSpectatorFactory.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+public class ConfigGeneratorClusterSpectatorFactory implements ClusterSpectatorFactory {
+
+  private final String uriPattern;
+
+  public ConfigGeneratorClusterSpectatorFactory(String uriPattern) {
+    this.uriPattern = uriPattern;
+  }
+
+  private String getConfigPostUri(String uriPattern, String clusterName) {
+    return uriPattern.replace("[PARTICIPANT_CLUSTERR]", clusterName);
```
Comment (gopalrajpurohit) [inline cluster_management/src/main/java/com/pinterest/rocksplicator/spectator/ConfigGeneratorClusterSpectatorFactory.java:12]: -> Seems spelling mistake `s/CLUSTERR/CLUSTER/g`

Note: Here we replace the '[PARTICIPANT_CLUSTER]' with the actual helix participant cluster that needs to be managed by this leader spectator.

File: cluster_management/src/main/java/com/pinterest/rocksplicator/DistributedSpectatorMain.java [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+import org.apache.helix.participant.StateMachineEngine;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+
+/**
+ * start spectator cluster manager controller
+ * spectator cluster manager controller runs in distributed mode:
+ * distributed mode: in this mode each spectator first joins as controller and participant into
```
[the remark this answers is not in the document]
Comment (gopalrajpurohit) [reply cluster_management/src/main/java/com/pinterest/rocksplicator/DistributedSpectatorMain.java:45]: Yes. Exactly same way we do on controller clusters.

File: cluster_management/src/main/java/com/pinterest/rocksplicator/DistributedSpectatorMain.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+      return cliParser.parse(cliOptions, cliArgs);
+    } catch (ParseException pe) {
+      logger.error("fail to parse command-line options. cliArgs: " + Arrays.toString(cliArgs), pe);
+      printUsage(cliOptions);
+      System.exit(1);
+    }
+    return null;
+  }
+
+  public static HelixManager startHelixController(
```
[the remark this answers is not in the document]
Comment (gopalrajpurohit) [reply cluster_management/src/main/java/com/pinterest/rocksplicator/DistributedSpectatorMain.java:150]: This part is actually a controller for the spectator-controller cluster itself... It is not the Spectator per se...

File: cluster_management/src/main/java/com/pinterest/rocksplicator/spectator/DistClusterSpectatorStateModel.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+import org.apache.helix.model.Message;
+import org.apache.helix.participant.AbstractHelixLeaderStandbyStateModel;
+import org.apache.helix.participant.statemachine.StateModelInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@StateModelInfo(initialState = "OFFLINE", states = {
+    "LEADER", "STANDBY"
+})
+public class DistClusterSpectatorStateModel extends AbstractHelixLeaderStandbyStateModel {
```
[the remark this answers is not in the document]
Comment (gopalrajpurohit) [reply cluster_management/src/main/java/com/pinterest/rocksplicator/spectator/DistClusterSpectatorStateModel.java:32]: This is the LeaderStandby model... This class is basically implementing the state model transitions.

File: cluster_management/src/main/java/com/pinterest/rocksplicator/spectator/ConfigGeneratorClusterSpectatorFactory.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+public class ConfigGeneratorClusterSpectatorFactory implements ClusterSpectatorFactory {
+
+  private final String uriPattern;
+
+  public ConfigGeneratorClusterSpectatorFactory(String uriPattern) {
+    this.uriPattern = uriPattern;
+  }
+
+  private String getConfigPostUri(String uriPattern, String clusterName) {
```
[the remark this answers is not in the document]
Comment (gopalrajpurohit) [reply cluster_management/src/main/java/com/pinterest/rocksplicator/spectator/ConfigGeneratorClusterSpectatorFactory.java:29]: trivia... Can remove it. yes.

File: cluster_management/src/main/java/com/pinterest/rocksplicator/DistributedSpectatorMain.java [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+ * start spectator cluster manager controller
+ * spectator cluster manager controller runs in distributed mode:
+ * distributed mode: in this mode each spectator first joins as controller and participant into
+ *   a special spectator CONTROLLER_CLUSTER. Leader election happens in this special
+ *   cluster. The one that becomes the leader controls all spectators (including itself
+ *   to become leaders of other clusters.
+ *
+ * The participant clusters are added as a resource into the special spectator controller cluster.
+ * The leader of the spectator controller cluster assign each resource (aka the helix participant
+ * clusters to be managed) one leader and multiple followers. The leader instance of this special
```
[the remark this answers is not in the document]
Comment (gopalrajpurohit) [reply cluster_management/src/main/java/com/pinterest/rocksplicator/DistributedSpectatorMain.java:52]: yes. will modify

**Review by rajathprasad — APPROVED** [review]
Looks good

**Review by ghanpipi — APPROVED** [review]
lgtm. i like the idea of using the similar controller resource management solution

**Pull request merged**